### PR TITLE
Track nullability of struct properties that are not auto-properties

### LINKED ISSUE #29619: Track nullability of struct properties that are not auto-properties
There should be no warning reported for `s.P.ToString()` or `s.Q.ToString()`. Currently a warning is reported for `P`.
```c#
using System.Runtime.CompilerServices;
#nullable enable
struct S
{
    private object? _p;
    internal object? P
    {
        get { return _p; }
        set { _p = value; }
    }
    internal object? Q
    {
        get;
        set;
    }
    static void M(S s)
    {
        s.P = new object();
        s.Q = new object();
        s.P.ToString(); // warning: maybe null
        s.Q.ToString(); // no warning
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #29619

### CODE REVIEW

**Comment by jcouv** [comment]
Added "Area-Compilers" label

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -1702,10 +1670,45 @@ private void SetNotNullResult(BoundExpression node)
...
+            }
+
+            if (type.SpecialType != SpecialType.None)
+            {
+                return true;
+            }
+
+            foreach (var member in ((NamedTypeSymbol)type).GetMembersUnordered())
+            {
+                switch (member.Kind)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >switch (member.Kind) [](start = 16, length = 20)

I am not sure I am following this logic. First of all, it looks order dependent. Depending on what members we get into first, we return different result. And we are not requesting members in any particular order. It would be good to add a comment describing what are we trying to figure out here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -29046,9 +29046,21 @@ struct S1
                 // (11,17): error CS0165: Use of unassigned local variable 'y1'
                 //         S1 z1 = y1;
                 Diagnostic(ErrorCode.ERR_UseDefViolation, "y1").WithArguments("y1").WithLocation(11, 17),
+                // (12,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// (12,14): warning CS8600: Converting null literal or possible null value to non-nullable type. [](start = 16, length = 96)

This and other warnings added in this unit-test look like a cascading diagnostics. Intentional?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Good catch, thanks.

---
In reply to: [284925656](https://github.com/dotnet/roslyn/pull/35765#discussion_r284925656) [](ancestors = 284925656)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Yes, this is a consequence of the change.

---
In reply to: [284929567](https://github.com/dotnet/roslyn/pull/35765#discussion_r284929567) [](ancestors = 284929567)

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 3]
```diff
@@ -313,7 +313,7 @@ internal partial class UnboundLambda
             ImmutableArray<string> names,
             bool isAsync,
             bool hasErrors = false)
-            : base(BoundKind.UnboundLambda, syntax, null, hasErrors || !types.IsDefault && types.Any(SymbolKind.ErrorType))
+            : base(BoundKind.UnboundLambda, syntax, null, hasErrors || !types.IsDefault && types.Any(t => t.Type?.Kind == SymbolKind.ErrorType))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: >t => t.Type?.Kind == SymbolKind.ErrorType [](start = 101, length = 41)

It is not obvious why this change is needed. Could you please elaborate?

Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: I removed two `SymbolExtensions.Any` methods. One was unused and the other was only used in this one location, so I replaced this use with the equivalent.

---
In reply to: [285212378](https://github.com/dotnet/roslyn/pull/35765#discussion_r285212378) [](ancestors = 285212378)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: > Yes, this is a consequence of the change.

Was it intentional?

---
In reply to: [284985684](https://github.com/dotnet/roslyn/pull/35765#discussion_r284985684) [](ancestors = 284985684,284929567)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -1702,10 +1664,52 @@ private void SetNotNullResult(BoundExpression node)
...
+
+            if (type.SpecialType != SpecialType.None)
+            {
+                return true;
+            }
+
+            var members = ((NamedTypeSymbol)type).GetMembersUnordered();
+
+            // EmptyStructTypeCache.IsEmptyStructType() returned true. If there
+            // are fields, those must be cyclic, so treat the type as empty.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >those must be cyclic [](start = 27, length = 20)

Is this accurate? Are all of them cyclic?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Not intentional, but the update behavior seems reasonable.

Logged [https://github.com/dotnet/roslyn/issues/35790](https://github.com/dotnet/roslyn/issues/35790) to consider removing the warning.

---
In reply to: [285213774](https://github.com/dotnet/roslyn/pull/35765#discussion_r285213774) [](ancestors = 285213774,284985684,284929567)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -49940,11 +49963,186 @@ static void F()
...
+@"class Program
+{
+    static void F(S s)
+    {
+        s.P = null;
+        s.P.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(new[] { source }, new[] { ref0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

Do we expect any diagnostics here? It looks like there is a difference in behavior by comparison to source only scenario. Do we understand why?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: No it's not accurate. At least one field is cyclic. I'll update in a subsequent PR.

---
In reply to: [285217976](https://github.com/dotnet/roslyn/pull/35765#discussion_r285217976) [](ancestors = 285217976)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This is the expected behavior. The difference is `EmptyStructTypeCache.IsEmptyStructType()` returns different values for the source and metadata cases.

---
In reply to: [285218545](https://github.com/dotnet/roslyn/pull/35765#discussion_r285218545) [](ancestors = 285218545)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: > The difference is EmptyStructTypeCache.IsEmptyStructType() returns different values for the source and metadata cases.

Is the cycle broken for source in some different way?

---
In reply to: [285219414](https://github.com/dotnet/roslyn/pull/35765#discussion_r285219414) [](ancestors = 285219414,285218545)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -49940,11 +49963,186 @@ static void F()
...
+  }
+}";
+            var ref0 = CompileIL(source0);
+            var source =
+@"class Program
+{
+    static void F(S s)
+    {
+        s.P = null;
+        s.P.ToString(); // 1
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >s.P.ToString(); // 1 [](start = 8, length = 20)

It feels inconsistent that tracking of properties in structures depend on presence of cycles in fields, but doesn't have the same dependency in classes.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: From source, cyclic fields are dropped from `EmptyStructTypeCache.GetStructInstanceFields()`.

---
In reply to: [285220520](https://github.com/dotnet/roslyn/pull/35765#discussion_r285220520) [](ancestors = 285220520,285219414,285218545)
